ddtrace/internal/_encoding.pyx

```diff
@@ -115,14 +132,152 @@ cdef inline int pack_text(msgpack_packer *pk, object text):
+
+        _id = self._next_id
+        PyDict_SetItem(self._table, string, PyLong_FromLong(_id))
+        self.insert(string)
+        self._next_id += 1
+        return _id
+
+    cpdef index(self, object string):
+        return self._index(string)
+
+    cdef reset(self):
+        PyDict_Clear(self._table)
+        self._next_id = 0
+        assert self._index("") == 0
+
+    def __len__(self):
+        return PyLong_FromLong(self._next_id)
+
+    def __contains__(self, object string):
+        return PyBool_FromLong(PyDict_Contains(self._table, string))
+
+
+cdef class ListStringTable(StringTable):
+    cdef list _list
+
+    def __init__(self):
+        self._list = []
+        super(ListStringTable, self).__init__()
+
+    cdef insert(self, object string):
+        PyList_Append(self._list, string)
+
+    def __iter__(self):
+        return iter(self._list)
+
+
+cdef class MsgpackStringTable(StringTable):
+    cdef msgpack_packer pk
+    cdef int max_size
+    cdef int _sp
+    cdef object _lock
+    cdef size_t _reset_size
+
+    def __init__(self, max_size):
+        self.pk.buf_size = min(max_size, 1 << 20)
+        self.pk.buf = <char*> PyMem_Malloc(self.pk.buf_size)
+        if self.pk.buf == NULL:
+            raise MemoryError("Unable to allocate internal buffer.")
+        self.max_size = max_size
+        self.pk.length = 6
+        self._sp = 0
+        self._lock = threading.Lock()
+        super(MsgpackStringTable, self).__init__()
+
+        assert self.index(ORIGIN_KEY) == 1
+        self._reset_size = self.pk.length
+
+    def __dealloc__(self):
+        PyMem_Free(self.pk.buf)
+        self.pk.buf = NULL
+
+    cdef insert(self, object string):
+        cdef int ret
+
+        ret = pack_text(&self.pk, string)
+        if ret != 0:
+            raise RuntimeError("Failed to add string to msgpack string table")
+
+    cdef savepoint(self):
+        self._sp = self.pk.length
+
+    cdef rollback(self):
+        if self._sp > 0:
+            self.pk.length = self._sp
+
+    cdef get_bytes(self):
+        cdef int ret;
+        cdef stdint.uint32_t l = self._next_id
+        cdef int offset = 6 - array_prefix_size(l)
+        cdef int old_pos = self.pk.length
+        
+        with self._lock:
+            # Update table size prefix
+            self.pk.length = offset
+            ret = msgpack_pack_array(&self.pk, l)
+            if ret:
+                return None
+            # Add root array size prefix
+            self.pk.length = offset = offset - 1
+            ret = msgpack_pack_array(&self.pk, 2)
+            if ret:
+                return None
+            self.pk.length = old_pos
+        
+        return PyBytes_FromStringAndSize(self.pk.buf + offset, self.pk.length - offset)
+
+    @property
+    def size(self):
+        return self.pk.length - MSGPACK_ARRAY_LENGTH_PREFIX_SIZE + array_prefix_size(self._next_id)
+
+    cdef append_raw(self, object src, Py_ssize_t size):
```

Comment:
```suggestion
    cdef append_raw(self, long src, Py_ssize_t size):
```

?

Reply:
Should help cut down on some of the Cython type checking + conversions.

Seems like we could even save going from long -> PyLong, and object -> PyLong

Reply:
In theory we don't need to pass a long, but a `char *`. However, for some reasons the wrong value was getting passed with that signature and that's why I have resorted to converting the address to a Python long and then back to an address 😢 .